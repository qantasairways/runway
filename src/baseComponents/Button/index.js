import React from 'react';
import PropTypes from 'prop-types';
import { withTheme } from 'emotion-theming';

import { CSS_SELECTOR_ACTIVE } from '../../constants/css';
import noop from '../../utils/noop';
import deepMerge from '../../utils/deepMerge';

const baseStyles = {
  button: (theme, props) => ({
    fontFamily: theme.fontFamily.main,
    fontSize: theme.fontSize.button,
    width: props.inline ? 'auto' : '100%',
    height: '3.125rem',
    borderRadius: theme.layout.borderRadius,
    borderWidth: 0,
    padding: '14px 24px 15px',
    fontWeight: 700,
    lineHeight: 1.56,
    letterSpacing: '0.1em',
    textTransform: 'uppercase',
    cursor: 'pointer',
    [CSS_SELECTOR_ACTIVE]: {
      transform: `translate(1px, 1px)`
    }
  })
  // Nested compoenets would go here for example
  // icon: theme => ({})
};

export function ThemedButton({ children, theme, themeStyles, ...buttonProps }) {
  return (
    <button
      type="button"
      {...buttonProps}
      css={deepMerge(
        baseStyles.button(theme, buttonProps),
        themeStyles.button(theme, buttonProps)
      )}
    >
      {children}
    </button>
  );
}

export const buttonPropTypes = {
  children: PropTypes.node,
  type: PropTypes.string,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  inline: PropTypes.bool
};

export const buttonDefaultProps = {
  children: 'button',
  type: 'button',
  onClick: noop,
  disabled: false,
  inline: false
};

ThemedButton.propTypes = {
  theme: PropTypes.shape().isRequired,
  themeStyles: PropTypes.shape({ button: PropTypes.func }),
  ...buttonPropTypes
};

ThemedButton.defaultProps = {
  children: 'button',
  themeStyles: { button: noop },
  ...buttonDefaultProps
};

export default withTheme(ThemedButton);
