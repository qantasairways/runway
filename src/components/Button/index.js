import { css } from '@emotion/core';
import React from 'react';
import PropTypes from 'prop-types';

import { CSS_SELECTOR_ACTIVE, CSS_SELECTOR_HOVER } from '../../constants/css';
import { fontFamily, fontSize, colours, buttons } from '../../theme/airways';

export const BUTTON_VARIANT_SOLID = 'solid';
export const BUTTON_VARIANT_HOLLOW = 'hollow';

const baseStyles = css({
  fontFamily: fontFamily.main,
  fontSize: fontSize.button,
  width: buttons.width,
  height: buttons.height,
  borderRadius: buttons.borderRadius,
  borderWidth: 0,
  padding: '14px 24px 15px',
  fontWeight: 700,
  lineHeight: 1.56,
  letterSpacing: '0.1em',
  textTransform: 'uppercase',
  cursor: 'pointer',
  [CSS_SELECTOR_ACTIVE]: {
    transform: `translate(${buttons.clickOffset}, ${buttons.clickOffset})`
  }
});

const solidStyles = css(
  {
    backgroundColor: colours.primary,
    color: colours.white,
    [CSS_SELECTOR_HOVER]: {
      backgroundColor: colours.primaryDark
    },
    [CSS_SELECTOR_ACTIVE]: {
      backgroundColor: colours.primaryDark
    }
  },
  baseStyles
);

const hollowStyles = css(
  {
    backgroundColor: colours.transparent,
    boxShadow: `0 0 0 2px ${colours.primary} inset`,
    color: colours.primary,
    fontWeight: 'normal',
    [CSS_SELECTOR_HOVER]: {
      boxShadow: `0 0 0 2px ${colours.primaryDark} inset`,
      color: colours.primaryDark
    },
    [CSS_SELECTOR_ACTIVE]: {
      boxShadow: `0 0 0 2px ${colours.primaryDark} inset`,
      color: colours.primaryDark
    }
  },
  baseStyles
);

const styleMap = {
  [BUTTON_VARIANT_SOLID]: solidStyles,
  [BUTTON_VARIANT_HOLLOW]: hollowStyles
};

const Button = ({
  className,
  label,
  variant,
  cssOverrides,
  ...buttonProps
}) => (
  <div className={className}>
    <button
      type="button"
      {...buttonProps}
      css={[styleMap[variant], ...cssOverrides]}
    >
      {label}
    </button>
  </div>
);

Button.propTypes = {
  label: PropTypes.string,
  variant: PropTypes.oneOf([BUTTON_VARIANT_HOLLOW, BUTTON_VARIANT_SOLID]),
  className: PropTypes.string,
  cssOverrides: PropTypes.arrayOf(PropTypes.object)
};

Button.defaultProps = {
  /** The text for the button */
  label: 'button',
  /** The design variant */
  variant: BUTTON_VARIANT_SOLID,
  /** Class name for the button */
  className: null,
  /** Css overrides for the button */
  cssOverrides: []
};

export default Button;
