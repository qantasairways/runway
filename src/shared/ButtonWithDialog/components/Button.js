import React from 'react';
import PropTypes from 'prop-types';
import { css } from '@emotion/core';

import noop from '../../../utils/noop';

import {
  fontSize,
  colours,
  letterSpacing,
  layout,
  fontFamily,
  fontWeight
} from '../../../theme/airways';

function buttonStyles({ hasMediumGreyback }) {
  return css({
    label: 'runway-popup-field__button',
    position: 'relative',
    height: '70px',
    width: '100%',
    maxWidth: '100%',
    border: 0,
    padding: `0 ${layout.gutter}`,
    cursor: 'pointer',
    backgroundColor: hasMediumGreyback
      ? colours.mediumGrey
      : colours.darkerGrey,
    color: colours.white,
    fontFamily: fontFamily.main,
    fontSize: fontSize.body,
    fontWeight: fontWeight.regular
  });
}

export function ButtonContent({ largeButtonValue, smallButtonValue }) {
  return (
    <div
      css={{
        width: '100%',
        minWidth: 0
      }}
    >
      <div
        css={{
          label: 'runway-dialog-button__value--large',
          fontSize: fontSize.large,
          letterSpacing: letterSpacing.small,
          lineHeight: 1
        }}
      >
        {largeButtonValue}
      </div>
      <div
        css={{
          label: 'runway-dialog-button__value--small',
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          whiteSpace: 'nowrap'
        }}
      >
        {smallButtonValue}
      </div>
    </div>
  );
}

ButtonContent.propTypes = {
  largeButtonValue: PropTypes.string,
  smallButtonValue: PropTypes.string
};

ButtonContent.defaultProps = {
  largeButtonValue: '',
  smallButtonValue: ''
};

function Button({
  renderButtonValue,
  iconLabelButtonValue,
  onClick,
  onBlur,
  setButtonRef,
  open,
  buttonLabel,
  placeHolder,
  Icon
}) {
  const buttonValue = renderButtonValue();

  const renderIconLabelPair = () => (
    <div
      css={{
        width: '100%',
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center'
      }}
    >
      <span
        css={{
          label: 'runway-popup-field__label',
          flex: '1',
          whiteSpace: 'nowrap',
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          textAlign: 'left'
        }}
      >
        {iconLabelButtonValue.label}
      </span>
      <span
        css={{
          width: layout.iconSize,
          height: layout.iconSize
        }}
      >
        <iconLabelButtonValue.icon
          css={{
            label: 'runway-popup-field__icon',
            right: '10px',
            fill: colours.white
          }}
          height="100%"
          width="100%"
        />
      </span>
    </div>
  );

  return (
    <button
      type="button"
      aria-haspopup="dialog"
      aria-expanded={open}
      onClick={onClick}
      onBlur={onBlur}
      ref={setButtonRef}
      css={
        iconLabelButtonValue
          ? buttonStyles({ hasMediumGreyback: true })
          : buttonStyles({ hasMediumGreyback: false })
      }
    >
      {iconLabelButtonValue ? (
        renderIconLabelPair()
      ) : (
        <div
          css={{
            label: 'runway-popup-field__label',
            top: '5px',
            left: '7px',
            color: 'white',
            textTransform: 'none',
            fontSize: fontSize.label,
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center'
          }}
        >
          <span
            css={{
              label: 'runway-popup-field__label',
              position: 'absolute',
              top: '5px',
              left: '7px',
              color: 'white',
              fontSize: fontSize.label
            }}
          >
            {buttonLabel}
          </span>
          {buttonValue}
          {!buttonValue && (
            <div
              css={{
                label: 'runway-popup-field__placeholder',
                color: colours.mediumDarkGrey,
                fontSize: fontSize.large,
                letterSpacing: letterSpacing.small
              }}
            >
              {placeHolder}
            </div>
          )}
          {!buttonValue && !!Icon && (
            <Icon
              css={{
                label: 'runway-popup-field__icon',
                position: 'absolute',
                right: '10px',
                fill: colours.mediumDarkGrey
              }}
              height={layout.iconSize}
              width={layout.iconSize}
            />
          )}
        </div>
      )}
    </button>
  );
}

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  renderButtonValue: PropTypes.func,
  onBlur: PropTypes.func,
  setButtonRef: PropTypes.func,
  open: PropTypes.bool,
  buttonLabel: PropTypes.string,
  placeHolder: PropTypes.string,
  Icon: PropTypes.func,
  iconLabelButtonValue: PropTypes.shape({
    label: PropTypes.string,
    icon: PropTypes.any
  })
};

Button.defaultProps = {
  onBlur: noop,
  setButtonRef: noop,
  renderButtonValue: () => null,
  open: false,
  buttonLabel: '',
  placeHolder: '',
  Icon: null,
  iconLabelButtonValue: null
};

export default Button;
