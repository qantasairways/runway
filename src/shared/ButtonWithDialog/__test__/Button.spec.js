import React from 'react';
import { shallow } from 'enzyme';
import Button, { ButtonContent } from '../components/Button';

const DummyIcon = () => <div>ICON</div>;

describe('ButtonContent', () => {
  let component;

  it('renders correctly with defaults', () => {
    component = shallow(<ButtonContent />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with props provided', () => {
    component = shallow(
      <ButtonContent largeValue="Large Value" smallValue="Small Value" />
    );

    expect(component).toMatchSnapshot();
  });
});

describe('Button', () => {
  let component;

  it('renders correctly with defaults', () => {
    component = shallow(<Button onClick={() => {}} setButtonRef={() => {}} />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with props provided', () => {
    component = shallow(
      <Button
        renderButtonvalue={() => 'button value'}
        open
        onClick={() => {}}
        onBlur={() => {}}
        setButtonRef={() => {}}
        buttonLabel="field label"
        placeHolder="placeholder"
        icon={DummyIcon}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with buttonContent custom element', () => {
    component = shallow(
      <Button
        renderButtonValue={() => <div>BUTTON CONTENT</div>}
        open
        onClick={() => {}}
        onBlur={() => {}}
        setButtonRef={() => {}}
        buttonLabel="field label"
        placeHolder="placeholder"
        icon={DummyIcon}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
