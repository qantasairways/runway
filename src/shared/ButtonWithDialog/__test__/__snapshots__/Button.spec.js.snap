// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly with buttonContent custom element 1`] = `
.emotion-2 {
  position: relative;
  height: 70px;
  width: 100%;
  max-width: 100%;
  border: 0;
  padding: 0 15px;
  cursor: pointer;
  background-color: #323232;
  color: #fff;
  font-family: "Ciutadella",sans-serif;
  font-size: 1.125rem;
  font-weight: 400;
}

.emotion-1 {
  top: 5px;
  left: 7px;
  color: white;
  text-transform: none;
  font-size: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  position: absolute;
  top: 5px;
  left: 7px;
  color: white;
  font-size: 1rem;
}

<Button
  Icon={null}
  buttonLabel="field label"
  icon={[Function]}
  iconLabelButtonValue={null}
  onBlur={[Function]}
  onClick={[Function]}
  open={true}
  placeHolder="placeholder"
  renderButtonValue={[Function]}
  setButtonRef={[Function]}
>
  <button
    aria-expanded={true}
    aria-haspopup="dialog"
    className="emotion-2"
    onBlur={[Function]}
    onClick={[Function]}
    type="button"
  >
    <div
      className="emotion-1"
    >
      <span
        className="emotion-0"
      >
        field label
      </span>
      <div>
        BUTTON CONTENT
      </div>
    </div>
  </button>
</Button>
`;

exports[`Button renders correctly with defaults 1`] = `
.emotion-3 {
  position: relative;
  height: 70px;
  width: 100%;
  max-width: 100%;
  border: 0;
  padding: 0 15px;
  cursor: pointer;
  background-color: #323232;
  color: #fff;
  font-family: "Ciutadella",sans-serif;
  font-size: 1.125rem;
  font-weight: 400;
}

.emotion-2 {
  top: 5px;
  left: 7px;
  color: white;
  text-transform: none;
  font-size: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  position: absolute;
  top: 5px;
  left: 7px;
  color: white;
  font-size: 1rem;
}

.emotion-1 {
  color: #888888;
  font-size: 1.75rem;
  -webkit-letter-spacing: -0.0625rem;
  -moz-letter-spacing: -0.0625rem;
  -ms-letter-spacing: -0.0625rem;
  letter-spacing: -0.0625rem;
}

<Button
  Icon={null}
  buttonLabel=""
  iconLabelButtonValue={null}
  onBlur={[Function]}
  onClick={[Function]}
  open={false}
  placeHolder=""
  renderButtonValue={[Function]}
  setButtonRef={[Function]}
>
  <button
    aria-expanded={false}
    aria-haspopup="dialog"
    className="emotion-3"
    onBlur={[Function]}
    onClick={[Function]}
    type="button"
  >
    <div
      className="emotion-2"
    >
      <span
        className="emotion-0"
      />
      <div
        className="emotion-1"
      />
    </div>
  </button>
</Button>
`;

exports[`Button renders correctly with props provided 1`] = `
.emotion-3 {
  position: relative;
  height: 70px;
  width: 100%;
  max-width: 100%;
  border: 0;
  padding: 0 15px;
  cursor: pointer;
  background-color: #323232;
  color: #fff;
  font-family: "Ciutadella",sans-serif;
  font-size: 1.125rem;
  font-weight: 400;
}

.emotion-2 {
  top: 5px;
  left: 7px;
  color: white;
  text-transform: none;
  font-size: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  position: absolute;
  top: 5px;
  left: 7px;
  color: white;
  font-size: 1rem;
}

.emotion-1 {
  color: #888888;
  font-size: 1.75rem;
  -webkit-letter-spacing: -0.0625rem;
  -moz-letter-spacing: -0.0625rem;
  -ms-letter-spacing: -0.0625rem;
  letter-spacing: -0.0625rem;
}

<Button
  Icon={null}
  buttonLabel="field label"
  icon={[Function]}
  iconLabelButtonValue={null}
  onBlur={[Function]}
  onClick={[Function]}
  open={true}
  placeHolder="placeholder"
  renderButtonValue={[Function]}
  renderButtonvalue={[Function]}
  setButtonRef={[Function]}
>
  <button
    aria-expanded={true}
    aria-haspopup="dialog"
    className="emotion-3"
    onBlur={[Function]}
    onClick={[Function]}
    type="button"
  >
    <div
      className="emotion-2"
    >
      <span
        className="emotion-0"
      >
        field label
      </span>
      <div
        className="emotion-1"
      >
        placeholder
      </div>
    </div>
  </button>
</Button>
`;

exports[`ButtonContent renders correctly with defaults 1`] = `
.emotion-2 {
  width: 100%;
  min-width: 0;
}

.emotion-0 {
  font-size: 1.75rem;
  -webkit-letter-spacing: -0.0625rem;
  -moz-letter-spacing: -0.0625rem;
  -ms-letter-spacing: -0.0625rem;
  letter-spacing: -0.0625rem;
  line-height: 1;
}

.emotion-1 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<ButtonContent
  largeButtonValue=""
  smallButtonValue=""
>
  <div
    className="emotion-2"
  >
    <div
      className="emotion-0"
    />
    <div
      className="emotion-1"
    />
  </div>
</ButtonContent>
`;

exports[`ButtonContent renders correctly with props provided 1`] = `
.emotion-2 {
  width: 100%;
  min-width: 0;
}

.emotion-0 {
  font-size: 1.75rem;
  -webkit-letter-spacing: -0.0625rem;
  -moz-letter-spacing: -0.0625rem;
  -ms-letter-spacing: -0.0625rem;
  letter-spacing: -0.0625rem;
  line-height: 1;
}

.emotion-1 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<ButtonContent
  largeButtonValue=""
  largeValue="Large Value"
  smallButtonValue=""
  smallValue="Small Value"
>
  <div
    className="emotion-2"
  >
    <div
      className="emotion-0"
    />
    <div
      className="emotion-1"
    />
  </div>
</ButtonContent>
`;
